### PR TITLE
[AJ-1279] Add support for date and date time data type conversions 

### PR DESCRIPTION
**Pull request by nawatts**
Stacked on #471. This adds support for date and date time types.

### CODE REVIEW

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java [hunk 1]
```diff
@@ -1087,6 +1082,41 @@ private String getPostgresTypeConversionExpression(
...
+      }
+    }
+
+    // Unable to cast dates or timestamps to numbers.
+    // Extract epoch from timestamp to get a number.
+    if ((dataType.getBaseType() == DataTypeMapping.DATE
+            || dataType.getBaseType() == DataTypeMapping.DATE_TIME)
+        && newDataType.getBaseType() == DataTypeMapping.NUMBER) {
+      if (dataType.isArrayType()) {
+        // TODO: confirm that this preserves order of array elements.
```
Comment (calypsomatic) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1104]: Is there a ticket for this TODO or is that as part of this PR?

Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java]: I was just in the process of removing that comment. It's from a previous version of these functions.

**Review by ashanhol — APPROVED** [review]
Nice job with thorough test cases. Glad to see number (array) to.. as well as all the Date (array) to.. Datetime (array) to...  👍

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java [hunk 2]
```diff
@@ -1072,12 +1072,7 @@ private String getPostgresTypeConversionExpression(
       String attribute, DataTypeMapping dataType, DataTypeMapping newDataType) {
     // Some data types are not yet supported.
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1073]: I think you might have ample coverage of this in the test that checks the DB outcomes...but I wonder if there is enough going on in this method with enough potential edge cases that it might be worth making the method non-private and annotate it as @VisibleForTesting.

```
@VisibleForTesting
String getPostgresType...
```

And then add some unit tests that verify some of the edgiest-case String construction.

File: service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+databaseChangeLog:
+  - changeSet:
+      id:  20240118_number_timestamp_array_conversions
+      author:  nwatts
+      changes:
+        -  sql:
+             dbms:  postgresql
+             splitStatements: false
+             sql:  |
+               create or replace function convert_array_of_numbers_to_timestamps(numeric[])
```
Comment (jladieu) [inline service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml:10]: This stuff is intense!  I'd love to see if you have a source recipe you followed for PG procs.

**Review by jladieu — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java (hunk 2 above)
Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1073]: Ooh, I like that idea.

File: service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml (hunk 3 above)
Comment (nawatts) [reply service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml:10]: This was all adapted from StackOverflow answers.

**Comment by nawatts** [comment]
Rebased after addressing feedback on #471.

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java [hunk 4]
```diff
@@ -1089,6 +1093,39 @@ private String getPostgresTypeConversionExpression(
 
     String expression = quote(SqlUtils.validateSqlString(attribute, ATTRIBUTE));
 
+    // Unable to cast numbers to dates or timestamps.
+    // Convert number to timestamp using to_timestamp.
+    if (dataType.getBaseType() == DataTypeMapping.NUMBER
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1098]: These newly introduced conditionals should use `.equals(` too :)

File: service/build.gradle [hunk 5]
```diff
@@ -229,6 +229,7 @@ jacocoTestReport {
 test {
     systemProperties['pact.rootDir'] = "$buildDir/pacts"
     systemProperties['pact.provider.version'] = "$project.version"
+    systemProperties['user.timezone'] = "UTC";
```
Comment (davidangb) [inline service/build.gradle:232]: :+1: for this code change - could you explain it to the team in a sprint review or team topics or some such?

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java [hunk 6]
```diff
@@ -1089,6 +1093,39 @@ private String getPostgresTypeConversionExpression(
...
+        }
+        expression = "(" + expression + ")";
+      } else {
+        expression = "extract(epoch from " + expression + ")";
+        // Truncate to an integer for dates.
+        if (dataType.getBaseType() == DataTypeMapping.DATE) {
+          expression += "::bigint";
+        }
+      }
+    }
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1127]: I think this needs Product input too - is this an appropriate conversion? Or should we disallow conversion between these datatypes? Any change in functionality here would of course extend to the `convert_array_of_timestamps_to_numbers` function

File: service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                 declare
+                    timestamps alias for $1;
+                    numbers numeric[];
+                  begin
+                    for i in array_lower(timestamps, 1)..array_upper(timestamps, 1) loop
+                      numbers[i] := extract(epoch from timestamps[i]);
+                    end loop;
+                    return numbers;
+                  end;
+               $$ language plpgsql stable;
```
Comment (davidangb) [inline service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml:37]: Could you declare these functions as [IMMUTABLE](https://www.postgresql.org/docs/14/xfunc-volatility.html), as a best practice? I believe these are immutable, correct? 

Also, for discussion: this creates these two functions inside the `public` schema. This is consistent with where we created the file, relation, array_of_file, and array_of_relation domains. But, should we create these functions in the `sys_wds` schema instead, and at some future point move our domains into `sys_wds` as well?

**Review by davidangb — COMMENTED** [review]
questions/thoughts inline, nice work on this!

File: service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml (hunk 7 above)
Comment (nawatts) [reply service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml:37]: Ooh, I learned about STABLE from StackOverflow, but did not know about IMMUTABLE.

The `sys_wds` schema question is beyond my Postgres knowledge.

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java (hunk 6 above)
Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1127]: I did take a decision here on making numbers represent seconds since epoch when converting between numbers and dates or date/times. From a programmer's perspective, I think that's a natural conversion, though one _could_ argue for milliseconds vs seconds, etc.

File: service/build.gradle (hunk 5 above)
Comment (yuliadub) [reply service/build.gradle:232]: yay UTC!

**Comment by nawatts** [comment]
Rebased after addressing feedback on #471

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java (hunk 4 above)
Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java]: Done in https://github.com/DataBiosphere/terra-workspace-data-service/pull/472/commits/2d11823bbcb838694d446c5077c959cc54c59f5f.

File: service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml (hunk 7 above)
Comment (nawatts) [reply service/src/main/resources/liquibase/changesets/20240118_number_timestamp_array_conversions.yaml]: Marked these functions immutable in https://github.com/DataBiosphere/terra-workspace-data-service/pull/472/commits/59d6c41ca72180399ef4a0776a22add7cc941012. Moved them to the `sys_wds` schema in https://github.com/DataBiosphere/terra-workspace-data-service/pull/472/commits/ce1a49d4328ca16757b4e0fd4e0d76d8ecf9d2ae.

File: service/build.gradle (hunk 5 above)
Comment (nawatts) [reply service/build.gradle:232]: When converting between datetimes and strings or between dates/datetimes and numbers, the timezone for the database session affects the result. We want the tests to pass with the same fixture data regardless of what timezone they're run in (UTC for CI, US Eastern or Pacific for developers' machines). To accomplish that, this  overrides the system timezone and to ensure tests are always run with the same timezone setting.

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java (hunk 2 above)
Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java]: Saving this for another PR since this one's already hundreds of lines.

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java [hunk 8]
```diff
@@ -1079,12 +1084,16 @@ public void updateAttributeDataType(
...
-                .formatted(attribute, newDataType.name()));
+      if (e.getRootCause() instanceof SQLException sqlEx && sqlEx.getSQLState() != null) {
+        if (expectedDataTypeConversionErrorCodes.contains(sqlEx.getSQLState())) {
+          throw new ConflictException(
+              "Unable to convert values for attribute %s to %s"
+                  .formatted(attribute, newDataType.name()));
+        } else {
+          LOGGER.warn(
+              "updateAttributeDataType: DataIntegrityViolationException with unexpected error code {}",
+              sqlEx.getSQLState());
```
Comment (nawatts) [inline service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1095]: Thought this might help us discover other data type conversion related error codes that we should possibly handle.

File: service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java (hunk 6 above)
Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1152]: @bethsheets

Comment (nawatts) [reply service/src/main/java/org/databiosphere/workspacedataservice/dao/RecordDao.java:1157]: Keeping this conversion per [Slack conversation](https://broadinstitute.slack.com/archives/C02SJQ0FKHQ/p1706544435880189?thread_ts=1706538890.918049&cid=C02SJQ0FKHQ) with @bethsheets.

**Review by yuliadub — APPROVED** [review]

**Review by davidangb — APPROVED** [review]
